src/prompt/functions/ai-chain.ts

```diff
@@ -0,0 +1,174 @@
+
+/**
+ * Creates a chain of chat completion calls that can be invoked as a single
+ * function. It is meant to simplify the process of resolving tool calls
+ * and optionally adding validation to the final result.
+ *
+ * The returned function will invoke the `chatModel` up to `maxCalls` times,
+ * resolving any tool calls to the included `functions` and retrying if
+ * necessary up to `maxRetries`.
+ *
+ * You must pass either `params.messages` or `prompt` to provide the initial
+ * messages to the chat completion model (passing both will throw an error).
+ *
+ * The chain ends when a non-tool call is returned, and the final result can
+ * optionally be validated against a Zod schema, which defaults to a `string`.
+ *
+ * To prevent possible infinite loops, the chain will throw an error if it
+ * exceeds `maxCalls` (`maxCalls` is expected to be >= `maxRetries`).
+ *
+ * Note that `createAIChain` takes in a `functions` array of `AIFunction`
+ * objects created by `createAIFunction`, as the two utility functions are
+ * meant to be used together.
+ */
+export function createAIChain<
+  Params extends Prompt.ChainParams = void,
+  Result extends Prompt.ChainResult = string
+>(
+  {
+    chatModel,
+    schema,
+    params,
+    prompt,
+    functions,
+    maxCalls = 5,
+    maxRetries = 3,
+    toolCallConcurrency = 8,
+  }: {
+    chatModel: Model.Chat.Model;
+    schema?: ZodType<Result>;
+    params?: Omit<
+      Prettify<Model.Chat.Run & Model.Chat.Config>,
+      'functions' | 'tools' | 'function_call' | 'tool_choice'
+    >;
+    prompt?: Prompt.Template<Params>;
+    functions?: Prompt.AIFunction[];
+    maxCalls?: number;
+    maxRetries?: number;
+    toolCallConcurrency?: number;
+  },
+  context?: Model.Ctx
+): Prompt.Chain<Params, Result> {
+  return async (promptArgs: Params): Promise<Result> => {
+    if (!params?.messages && !prompt) {
+      throw new Error('Either params.messages or prompt must be provided');
+    }
+
+    if (params?.messages && prompt) {
+      throw new Error('Only one of params.messages or prompt may be provided');
+    }
+
+    const messages =
+      params?.messages ?? (await Promise.resolve(prompt!(promptArgs)));
+    const tools = functions?.map((func) => ({
+      type: 'function' as const,
+      function: func.spec,
+    }));
+
+    let numCalls = 0;
+    let numErrors = 0;
+
+    while (numCalls < maxCalls) {
+      ++numCalls;
+      const response = await chatModel.run(
+        {
+          ...params,
+          messages,
+          tools,
+        },
+        context
+      );
+
+      const { message } = response;
+      try {
+        if (Msg.isToolCall(message)) {
+          messages.push(message);
+
+          if (!functions) {
+            throw new AbortError('No functions provided to handle tool call');
+          }
+
+          await pMap(
+            message.tool_calls,
+            async (toolCall) => {
+              const func = functions?.find(
```

Comment:
is it worth making a map of the functions?

Reply:
good call

Reply:
done